pandas/core/generic.py

```diff
@@ -6308,11 +6308,23 @@ def fillna(
         else:
             if self.ndim == 1:
                 if isinstance(value, (dict, ABCSeries)):
-                    value = create_series_with_explicit_dtype(
+                    value_map = create_series_with_explicit_dtype(
                         value, dtype_if_empty=object
                     )
-                    value = value.reindex(self.index, copy=False)
-                    value = value._values
+                    # GH#40498 objects can have multiple types of missing values which
+                    # should not be modified unless specified.  Add special casing to
+                    # minimize performance decrease on other data types where this is
+                    # not required.
+                    if is_object_dtype(self.dtype):
+                        value = self.copy()
+                        modification_index = value.index.intersection(value_map.index)
+                        if not modification_index.empty:
+                            value.loc[modification_index] = value_map[
+                                modification_index
+                            ]
+                    else:
+                        value = value_map.reindex(self.index, copy=False)
```

Comment:
why can't you just set `copy=True` for object type?

Reply:
This is the original logic that was present for all value dictionaries before I added the separate object type conditional.  

I'm not sure why they chose to explicitly set copy=False since in both the old and new code we're replacing value with the reindexed version and the original value is not used anywhere else.  In addition, unless every index is a key in the value dictionary this will have no effect because - "A new object is produced unless the new index is equivalent to the current one and copy=False".

Will remove the unnecessary argument and return to the default since there's no clear value for it.